Start removing old Python STs

caseydavenport opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

The impetus here is that SemaphoreCI is upgrading to a new kernel that doesn't support cgroupv1, required for the version of docker in docker used in these tests to function.

We've wanted to clean these up for a while anyway, so this is good motivation. Some of them are already redundant with other UTs/FVs - these can just be deleted.

Some others should be moved to other test suites (e.g., Felix FVs or e2e)

## Related issues/PRs

<!-- If appropriate, include a link to the issue this fixes.
fixes <ISSUE LINK>

If appropriate, add links to any number of PRs documented by this PR
documents <PR LINK>
-->

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

stevegaossou commented on Makefile:
```diff
@@ -164,15 +164,15 @@ image:
 # using a local kind cluster.
 ###############################################################################
-E2E_FOCUS ?= "sig-network.*Conformance|sig-calico.*Conformance"
+E2E_FOCUS ?= "sig-network.*Conformance|sig-calico.*Conformance|BGP"
```
Any reason to have the new BGP tests be in the default focus? Were the old python tests run as part of conformance?

caseydavenport replied:
This isn't really meant to define what Conformance is, it's just meant to define what tests we run as part of CI.

I should probably rename the Semaphore Job.

stevegaossou commented:
> Some others should be moved to other test suites (e.g., Felix FVs or e2e)

@caseydavenport could you confirm that you basically removed all of the problematic tests in this one PR (i.e. the DinD dependent tests). I'm assuming that is true judging by the volume of Python test code removed. 

And you have replaced some of them here. But there are a few remaining tests that got removed here that will need to be replaced with a Go-equivalent test in the near future. 

Is that correct?

caseydavenport commented:
@stevegaossou this PR doesn't get rid of every single DinD test - there are still a few more that I was going to do in a follow-on. 

The tests that I have removed fall into one of a few buckets (in order of prevalence in this PR):
- Replaced by a Go test in this PR.
- Already covered by another UT or e2e, so we don't need to replace.
- Replaced by a Python k8st in this PR.
- Low value test that doesn't need immediate replacement (but I have tracked in my doc)

stevegaossou commented on node/tests/k8st/tests/test_health.py:
```diff
@@ -0,0 +1,99 @@
+_log.setLevel(logging.DEBUG)
+
+
+class TestReadiness(TestBase):
```
@caseydavenport is the expectation that this new test logic written in Python will be able to live with the remaining Python tests for as long as we want to keep them around? 

Meaning, there isn't any reason in the near future to move these over to Golang? 

Was the reason to leave them in Python form simply a matter of being pragmatic? I.e. it would have been more overhead to convert them over because of surrounding utility functions etc (or that it makes more sense to be grouped with other existing Python tests).

caseydavenport commented on libcalico-go/lib/backend/k8s/resources/node.go:
```diff
@@ -49,13 +49,14 @@ const (
 	nodeWireguardPublicKeyV6Annotation    = "projectcalico.org/WireguardPublicKeyV6"
 	nodeInterfacesAnnotation              = "projectcalico.org/Interfaces"
+
+	RouteReflectorClusterIDAnnotation = "projectcalico.org/RouteReflectorClusterID"
```
Yeah, I use it in some of the new BGP tests.

nelljerram commented on e2e/pkg/tests/bgp/utils.go:
```diff
@@ -0,0 +1,107 @@
+func setNodeAsRouteReflector(cli ctrlclient.Client, rrNode *corev1.Node, rrClusterID string) {
+	By(fmt.Sprintf("Using node %s as a route reflector", rrNode.Name))
+	prePatch := ctrlclient.MergeFrom(rrNode.DeepCopy())
+	rrNode.Labels[resources.RouteReflectorClusterIDAnnotation] = rrClusterID
```
Do we need to set this as a label, as well as as an annotation?

nelljerram commented on e2e/pkg/tests/networking/ipip.go:
```diff
@@ -0,0 +1,250 @@
+			// - Create a client pod and assert that it can connect to the service.
+			By(fmt.Sprintf("Creating server pod in namespace %s", f.Namespace.Name))
+			// Use cuztomizers to ensure pods use the test IP pool and avoid landing on the same node.
+			cusomtizer := conncheck.CombineCustomizers(
```
```suggestion
			customizer := conncheck.CombineCustomizers(
```

nelljerram commented on node/tests/st/bgp/test_ipip.py:
```diff
@@ -1,626 +0,0 @@
-                                          host.execute("ip r"))
-                        retry_until_success(check, retries=10)
-
-                    # Check that routes are not flapping, on either host, by
```
This is something I remember as having been important once - i.e. we had a bug where the symptom was a persistent flapping with tunnel and non-tunnel routes.

caseydavenport replied:
Yeah, I think a more detailed ip monitor test would be good to have. I've added a note to the doc to re-implement this, likely in the k8st suite.

nelljerram commented on node/tests/st/bgp/test_route_reflector_cluster.py:
```diff
@@ -1,132 +0,0 @@
-            # with a different set of redundant route reflectors.
-            for host in [host1, host2, host3]:
-                for rr in rrc.get_redundancy_group():
-                    create_bgp_peer(host, "node", rr.ip, 64513, metadata={'name': host.name + rr.name.lower()})
```
Do we have a replacement for this?  IIRC the e2e test does not use multiple RR cluster IDs.

nelljerram commented on node/tests/st/bgp/test_switch_route_reflector.py:
```diff
@@ -1,145 +0,0 @@
-    def _test_switch_route_reflector(self, backend='bird', bgpconfig_as_num=64514, peer_as_num=64514):
-        """
-        Test that switching from node-to-node full mesh to route reflectors doesn't disrupt dataplane traffic if done as per
-        https://projectcalico.docs.tigera.io/networking/bgp#change-from-node-to-node-mesh-to-route-reflectors-without-any-traffic-disruption
```
Need replacement for this one?  Seems to require more than the current e2e connchecker can do.

caseydavenport replied:
Yeah, the continuous traffic checking is beyond what it can do - I was intending to add an enhancement for continuous checking in a follow-on (maybe using iperf, or just a looping ping of some sort).

I'll see about just adding it to this batch, though.

caseydavenport replied:
https://tigera.atlassian.net/browse/CORE-12003

I also have a follow-on PR in the works but think it is large enough to warrant its own PR.

nelljerram reviewed (approved):
Have approved so as not to block, but...
1. Have noted a few nits and concerns - happy to leave those to your judgement.
2. I feel like a lot of testing is being deleted that is still useful and not yet replaced in any other form - which is fine for this PR, so long as we have good tracking for recreating that in further work.  Please could you consider all the cases that I've pointed out and create detailed tracking for recreating those, if it doesn't already exist?  (I think we need something a bit more precise than the existing analysis doc; perhaps JIRA tickets.)

caseydavenport commented on e2e/pkg/tests/bgp/route_reflector.go:
```diff
@@ -0,0 +1,224 @@
+			By("Removing one of the route reflectors")
+			setNodeAsNotRouteReflector(cli, &nodes.Items[1])
+
+			// Verify c nnectivity is maintained.
```
that's a strange one

caseydavenport commented on node/tests/st/bgp/test_ipip.py:
```diff
@@ -1,626 +0,0 @@
-    @parameterized.expand([
-        ('bird',),
-    ])
-    def test_issue_1584(self, backend):
```
Yeah, I haven't recreated this one entirely, although I did cover some of it with a test that swaps IPIP mode and checks for address assignment and routing updates. 

I think the main thing that is missing from this test was the "cold start" aspect, which I felt wasn't really necessary to test explicitly in an e2e any more.

caseydavenport commented on node/tests/st/bgp/test_ipip.py:
```diff
@@ -1,626 +0,0 @@
-        (False,),
-        (True,),
-    ])
-    def test_gce(self, with_ipip, backend='bird'):
```
I left this one out because we now have e2e suites that actually run regularly in e2e in various configurations, so I didn't think this was necessary any more. I'll add it to the doc, not sure how I missed it there!

caseydavenport commented on node/tests/st/bgp/test_node_peers.py:
```diff
@@ -1,80 +0,0 @@
-            check_bird_status(host2, [("node specific", host1.ip, "Established")])
-
-    @attr('slow')
-    def test_bird_node_peers(self):
```
Ah yeah. I think I had considered this covered by the BGPPeer e2e I wrote that uses selectors, which admittedly is subtly different. Although, I do think it's not really worth writing e2e tests for all of the different ways to select a node - especially since we have confd tests that cover this to ensure our config generation is correct: https://github.com/caseydavenport/calico/blob/ae5d06492d3921a6c224366784e218a39f776c75/confd/tests/mock_data/calicoctl/explicit_peering/specific_node/input.yaml#L9-L16

I'll add a new test for this regardless.

caseydavenport commented on node/tests/st/bgp/test_update_ip_addr.py:
```diff
@@ -1,68 +0,0 @@
-    def test_update_ip_address(self):
-        """
-        Test updating the IP address automatically updates and fixes the
-        Bird BGP config.
```
I believe this to be suitably covered by existing unit tests in libcalico and confd.

caseydavenport commented on node/tests/st/bgp/test_route_reflector_cluster.py:
```diff
@@ -1,132 +0,0 @@
-
-    def test_route_reflector_cluster_resilience(self):
-        """
-        Runs a cluster of route reflectors, brings one node down, and ensures that traffic still flows
```
It doesn't - I don't think there's a great way to do this in e2e, but I'll see what I can do. Might be enough to just delete the calico/node pod on one of the nodes and verify consistent connectivity.
